### PR TITLE
ci: add `golangci` linting

### PR DESCRIPTION
**Pull request by austinvalle**
contributes to:
- https://github.com/hashicorp/terraform-providers-devex-internal/issues/102
- https://github.com/hashicorp/terraform-providers-devex-internal/issues/83

### Notes
- I skipped the tests written with `sdk2`

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    - durationcheck
+    - errcheck
+    - exportloopref
+    - forcetypeassert
+    - gofmt
+    - gosimple
+    - ineffassign
+    - makezero
+    - misspell
+    - nilerr
```
Comment (austinvalle) [inline .golangci.yml:17]: Do we usually skip `paralleltest` for providers?

Comment (bflad) [reply .golangci.yml:17]: We can discuss this one out of band -- providers tend to implement [`resource.ParallelTest()`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource#ParallelTest), which all its doing under the hood is call `t.Parallel()` and [`resource.Test()`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource#Test). I believe these would all show as false positives.

File: internal/protocolprovider/data_time_test.go [hunk 2]
```diff
@@ -9,8 +9,11 @@ import (
 )
 
 func TestAccDataSourceTime(t *testing.T) {
+	t.Parallel()
+
 	resource.UnitTest(t, resource.TestCase{
 		ProtoV5ProviderFactories: map[string]func() (tfprotov5.ProviderServer, error){
+			//nolint:unparam // False positive in unparam related to map: https://github.com/mvdan/unparam/issues/40
```
Comment (bflad) [inline internal/protocolprovider/data_time_test.go:16]: Liked and subscribed!

File: internal/protocolprovider/server.go [hunk 3]
```diff
@@ -30,13 +30,13 @@ func (s server) GetProviderSchema(ctx context.Context, req *tfprotov5.GetProvide
...
 
-func (s server) PrepareProviderConfig(ctx context.Context, req *tfprotov5.PrepareProviderConfigRequest) (*tfprotov5.PrepareProviderConfigResponse, error) {
+func (s *server) PrepareProviderConfig(ctx context.Context, req *tfprotov5.PrepareProviderConfigRequest) (*tfprotov5.PrepareProviderConfigResponse, error) {
 	return &tfprotov5.PrepareProviderConfigResponse{
 		PreparedConfig: req.Config,
 	}, nil
 }
 
-func (s server) ConfigureProvider(ctx context.Context, req *tfprotov5.ConfigureProviderRequest) (*tfprotov5.ConfigureProviderResponse, error) {
+func (s *server) ConfigureProvider(ctx context.Context, req *tfprotov5.ConfigureProviderRequest) (*tfprotov5.ConfigureProviderResponse, error) {
```
Comment (bflad) [inline internal/protocolprovider/server.go:39]: 🐛 fix due to `s.client =` below  🎉

**Review by bflad — APPROVED** [review]
Looks good to me 🚀

File: internal/protocolprovider/data_time_test.go [hunk 4]
```diff
@@ -9,8 +9,11 @@ import (
 )
 
 func TestAccDataSourceTime(t *testing.T) {
+	t.Parallel()
```
Comment (austinvalle) [inline internal/protocolprovider/data_time_test.go:12]: Whoops, one slipped my commit, removing 🔥

File: .golangci.yml (hunk 1 above)
Comment (austinvalle) [reply .golangci.yml:17]: Cool, I'll leave it off the providers then 👍🏻

**Pull request merged**